File: python/vedant_bhatnagar/DevsnestProblems_dailyUpdated/SetMatrixZeroes.py
def solve(a):
    m, n = len(a), len(a[0])
    r,c = False, False
    for j in range(n):
        if(a[0][j] == 0):
            r = True
            break
    for i in range(m):
        if(a[i][0] == 0):
            c = True
            break

    for i in range(m):
        for j in range(n):
            if a[i][j] == 0:
                a[i][0] = 0
                a[0][j] = 0

    for i in range(1,m):
        if(a[i][0] == 0):
            for j in range(1,n):
                a[i][j] = 0

    for j in range(1,n):
        if(a[0][j] == 0):
            for i in range(1,m):
                a[i][j] = 0

    if r:
        for j in range(n):
            a[0][j] = 0
    if c:
        for i in range(m):
            a[i][0] = 0

    return a


#using set approach

    row = set()
    col = set()
    m, n = len(a), len(a[0])
    for i in range(m):
        for j in range(n):
            if a[i][j] == 0:
                row.add(i)
                col.add(j)

    for i in row:
        for j in range(0,n):
            a[i][j] = 0

    for j in col:
        for i in range(0,m):
            a[i][j] = 0

    return a

File: java/sky/LeetCode/aug-30-2022/src/set_matrix_zeroes.java
//link: https://leetcode.com/problems/set-matrix-zeroes/submissions/
//Approach : Most optimized approach TC: O(m*n), SC: O(1)

class Solution {
    public void setZeroes(int[][] matrix) {
        int n = matrix.length;
        int m = matrix[0].length;
        boolean firstColumn = false, firstRow = false;
        //marking rows and cols with 0
        for(int i = 0; i<n; i++){
            for(int j = 0; j<m; j++){
                if(matrix[i][j] == 0){
                    if(i == 0)  firstRow = true;
                    if(j == 0)  firstColumn = true;
                    matrix[0][j] = 0;
                    matrix[i][0] = 0;
                }
            }
        }

        //making whole row and col 0 which are marked 0
        //cols
        for(int i = 1; i<n; i++){
            if(matrix[i][0] == 0){
                for(int j = 0; j<m; j++){
                    matrix[i][j] = 0;
                }
            }
        }
        //rows
        for(int j = 1; j<m; j++){
            if(matrix[0][j] == 0){
                for(int i = 0; i<n; i++){
                    matrix[i][j] = 0;
                }
            }
        }

        if(firstColumn){
            for(int i = 0; i<n; i++){
                matrix[i][0] = 0;
            }
        }
        if(firstRow){
            for(int j = 0; j<m; j++){
                matrix[0][j] = 0;
            }
        }

    }
}